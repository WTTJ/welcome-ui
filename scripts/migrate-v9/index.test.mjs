import fs from 'fs'
import os from 'os'
import path from 'path'

import { describe, expect, it } from 'vitest'

import { findAllComponentUsages } from './index.mjs'
import { processComponents } from './process-components.mjs'
import { transform, transformSpecificValue, transformValue } from './transform.mjs'

describe('upgrade-v9 migration script', () => {
  describe('processComponents (integration)', () => {
    it('migrates fixtures/Component1.tsx as expected', async () => {
      // Setup temp dir and copy fixture
      const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'wui-upgrade-test-'))
      const srcFile = path.join(__dirname, './fixtures/Component1.tsx')
      const destFile = path.join(tmpDir, 'Component1.tsx')
      fs.copyFileSync(srcFile, destFile)

      // Find components in the temp file
      const components = findAllComponentUsages(tmpDir)
      // Run migration (shouldReplace = true)
      await processComponents(components, true)

      // Read migrated file
      const migrated = fs.readFileSync(destFile, 'utf8')

      // Assert output (snapshot or inline string)
      expect(migrated).toMatchSnapshot()
    })

    it('migrates fixtures/Component2.tsx as expected', async () => {
      // Setup temp dir and copy fixture
      const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'wui-upgrade-test-'))
      const srcFile = path.join(__dirname, './fixtures/Component2.tsx')
      const destFile = path.join(tmpDir, 'Component2.tsx')
      fs.copyFileSync(srcFile, destFile)

      // Find components in the temp file
      const components = findAllComponentUsages(tmpDir)
      // Run migration (shouldReplace = true)
      await processComponents(components, true)

      // Read migrated file
      const migrated = fs.readFileSync(destFile, 'utf8')

      // Assert output (snapshot or inline string)
      expect(migrated).toMatchSnapshot()
    })
  })

  describe('findAllComponentUsages', () => {
    it('finds Box, Flex, Grid, Stack components in a directory', () => {
      // Setup: create a mock directory structure in memory or use a temp dir
      // For demonstration, just check that the function runs and returns an array
      const result = findAllComponentUsages(path.join(__dirname, './fixtures'))
      expect(Array.isArray(result)).toBe(true)
    })
  })

  describe('transformValue', () => {
    it('transforms known keys using valueMap', () => {
      expect(transformValue('align', 'center')).toContain('items')
      expect(transformValue('backgroundColor', 'red')).toContain('bg')
    })
    it('returns undefined for unknown keys', () => {
      expect(transformValue('unknownKey', 'foo')).toBeUndefined()
    })
  })

  describe('transform', () => {
    it('handles 100% as full', () => {
      expect(transform('w', '100%')).toBe('w-full')
    })
    it('handles numeric values', () => {
      expect(transform('w', '16')).toContain('w-[')
    })
    it('handles 0 as zero class', () => {
      expect(transform('m', '0')).toBe('m-0')
    })
  })

  describe('transformSpecificValue', () => {
    it('handles grid-cols special cases', () => {
      expect(transformSpecificValue('grid-cols-1fr')).toBe('grid-cols-1')
      expect(transformSpecificValue('grid-cols-1fr 1fr')).toBe('grid-cols-2')
    })
    it('returns value for non-grid-cols', () => {
      expect(transformSpecificValue('foo')).toBe('foo')
    })
  })
})
