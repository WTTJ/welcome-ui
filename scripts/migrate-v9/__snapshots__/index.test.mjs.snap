// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`upgrade-v9 migration script > processComponents (integration) > migrates fixtures/Component.tsx as expected 1`] = `
"import { Alert } from '../../../lib/src/old/components/Alert'
import { Box } from '../../../lib/src/old/components/Box'
import { IconsFont } from '../../../lib/src/old/components/IconsFont'
import { Link } from '../../../lib/src/old/components/Link'
import { Text } from '../../../lib/src/old/components/Text'

const FormattedMessage = ({
  defaultMessage,
  id,
}: {
  defaultMessage: string
  id: string
}): React.ReactNode => <span id={id}>{defaultMessage}</span>

const signinPath = '/signin'

// Usage examples
export const Example = () => (
  <>
    <div className="flex h-full">
      <aside className="bg-beige-20 flex-col grow h-max-content items-center justify-center lg:px-3xl md:px-xl max-w-[30rem] md:flex none min-h-full pb-xl w-40%">
        <Link to={signinPath}>
          <IconsFont.Add h={58} />
          <Text as="h1" mt="md" variant="h4">
            <FormattedMessage defaultMessage="Terms of use" id="EBDrtq" />
          </Text>
        </Link>
      </aside>
    </div>
    <Alert isFullWidth mx="auto" variant="danger">
      <Alert.Title mb="sm">
        <FormattedMessage defaultMessage="Please sign in with provider" id="5nmpcx" />
      </Alert.Title>
      <span>
        <FormattedMessage
          defaultMessage="Your organization enforces provider sign-in. Contact your administrator if you need help."
          id="TzKv+W"
        />
      </span>
    </Alert>
  </>
)
"
`;
